imcache.go

```diff
@@ -311,31 +315,31 @@ func (c *Cache[K, V]) GetOrSet(key K, val V, exp Expiration) (value V, present b
 // If you want to add or replace an entry, use the Set method instead.
 func (c *Cache[K, V]) Replace(key K, val V, exp Expiration) (present bool) {
 	now := time.Now()
-	newEntry := entry[K, V]{val: val}
+	newEntry := entry[K, V]{key: key, val: val}
 	exp.apply(&newEntry.exp)
 	newEntry.SetDefaultOrNothing(now, c.defaultExp, c.sliding)
 	c.mu.Lock()
 	if c.closed {
 		c.mu.Unlock()
 		return false
 	}
-	currentEntry, ok := c.m[key]
+	currentNode, ok := c.m[key]
 	if !ok {
 		c.mu.Unlock()
 		return false
 	}
-	newEntry.node = currentEntry.node
-	c.queue.Remove(currentEntry.node)
+	currentEntry := currentNode.entry()
 	if currentEntry.HasExpired(now) {
+		c.queue.remove(currentNode)
 		delete(c.m, key)
 		c.mu.Unlock()
 		if c.onEviction != nil {
 			go c.onEviction(key, currentEntry.val, EvictionReasonExpired)
 		}
 		return false
 	}
-	c.queue.Add(newEntry.node)
-	c.m[key] = newEntry
+	c.queue.touch(currentNode)
+	currentNode.setEntry(newEntry)
```

Comment:
Reverse the order of `queue.touch` and `node.setEntry`. There might be an eviction policy relying on the entry data (e.g. expiration).